### PR TITLE
Add flex attention backend

### PR DESCRIPTION
**Pull request by Leymore**
This PR implements NATTEN using [FlexAttention](https://pytorch.org/blog/flexattention/).

The provided interfaces include `flex_na1d`, `flex_na2d`, and `flex_na3d`, which can replace `na1d`, `na2d`, and `na3d`, but with `kernel_size`, `dilation` and `is_casual` support only.

Since `flex_attention` requires kernel compilation, the first run may take longer than usual.

## Usage

```python
import torch
from natten.flex import flex_na2d

batch_size, image_height, image_width, num_head, head_dim = 1, 64, 64, 8, 64
query = torch.randn(batch_size, image_height, image_width, num_head, head_dim, device='cuda')
key = torch.randn(batch_size, image_height, image_width, num_head, head_dim, device='cuda')
value = torch.randn(batch_size, image_height, image_width, num_head, head_dim, device='cuda')

output = flex_na2d(query, key, value, kernel_size=11, dilation=1, is_causal=False)
```

## Bug

- When `head_dim=64`, `num_tokens<=32`, (`torch.float16` or `torch.bfloat16`) and `torch.compile(flex_attention)` is used, `flex_natten` produces NaN gradients during backward propagation, even though the forward results remain correct.
    - This is a bug in the `flex_attention` kernel, but the exact trigger conditions are not fully identified. This issue will be reported to the PyTorch team.

## TODO

- [ ] Optimize dilation implementation.

### CODE REVIEW

File: src/natten/flex.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+                mask = ((kernel_center_x - kv_idx[i]).abs() <= dilate_kernel // 2) & (
+                    (q_idx[i] % dilation[i]) == (kv_idx[i] % dilation[i])
+                )
+
+            masks.append(mask)
+
+        return functools.reduce(lambda x, y: x & y, masks)  # type: ignore
+
+    seq_length = functools.reduce(lambda x, y: x * y, image_shape)
+    block_mask = create_block_mask(natten_mask_mod, 1, 1, seq_length, seq_length)  # type: ignore
```
Comment (alihassanijr) [inline src/natten/flex.py:81]: Can we set batch and heads to `None` as done in the blog post? NATTEN masks are also independent of batch/heads. 

```suggestion
    block_mask = create_block_mask(natten_mask_mod, B=None, H=None, Q_LEN=seq_length, KV_LEN=seq_length)  # type: ignore
```

File: src/natten/flex.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+@functools.lru_cache(maxsize=None)
+def get_block_mask(
+    num_dimension: int,
+    image_shape: Tuple[int],
+    kernel_size: Tuple[int],
+    dilation: Tuple[int],
+    is_causal: Tuple[bool],
+):
+
+    def get_location_1d(idx: IntTensor) -> Tuple[IntTensor]:
```
Comment (alihassanijr) [inline src/natten/flex.py:33]:
```suggestion
    def index_to_coord_1d(idx: IntTensor) -> Tuple[IntTensor]:
```

File: src/natten/flex.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        return (idx,)
+
+    def get_location_2d(idx: IntTensor) -> Tuple[IntTensor, IntTensor]:
```
Comment (alihassanijr) [inline src/natten/flex.py:36]:
```suggestion
    def index_to_coord_2d(idx: IntTensor) -> Tuple[IntTensor, IntTensor]:
```

File: src/natten/flex.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        return (idx // image_shape[1], idx % image_shape[1])  # type: ignore
+
+    def get_location_3d(idx: IntTensor) -> Tuple[IntTensor, IntTensor, IntTensor]:
```
Comment (alihassanijr) [inline src/natten/flex.py:39]:
```suggestion
    def index_to_coord_3d(idx: IntTensor) -> Tuple[IntTensor, IntTensor, IntTensor]:
```

File: src/natten/flex.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        return (
+            idx // image_shape[2] // image_shape[1],  # type: ignore
+            (idx // image_shape[2]) % image_shape[1],  # type: ignore
+            idx % image_shape[2],  # type: ignore
+        )
+
+    get_location = {
+        1: get_location_1d,
+        2: get_location_2d,
+        3: get_location_3d,
```
Comment (alihassanijr) [inline src/natten/flex.py:49]:
```suggestion
    index_to_coord = {
        1: index_to_coord_1d,
        2: index_to_coord_2d,
        3: index_to_coord_3d,
```

File: src/natten/flex.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    }[num_dimension]
+
+    def natten_mask_mod(
+        b: IntTensor, h: IntTensor, q_idx: IntTensor, kv_idx: IntTensor
+    ) -> BoolTensor:
+        q_idx = get_location(q_idx)  # type: ignore
+        kv_idx = get_location(kv_idx)  # type: ignore
```
Comment (alihassanijr) [inline src/natten/flex.py:56]:
```suggestion
        q_coord = index_to_coord(q_idx)  # type: ignore
        kv_coord = index_to_coord(kv_idx)  # type: ignore
```

File: src/natten/flex.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+                    & ((q_idx[i] % dilation[i]) == (kv_idx[i] % dilation[i]))
+                )
+            else:
+                kernel_center_x = q_idx[i].clamp(
+                    (dilate_kernel - 1) // 2,
+                    (image_shape[i] - 1) - (dilate_kernel - 1) // 2,
+                )
+                mask = ((kernel_center_x - kv_idx[i]).abs() <= dilate_kernel // 2) & (
+                    (q_idx[i] % dilation[i]) == (kv_idx[i] % dilation[i])
+                )
```
Comment (alihassanijr) [inline src/natten/flex.py:74]:
```suggestion
            kernel_times_dilation = kernel_size[i] * dilation[i]
            if is_causal[i]:
                mask = (
                    (q_coord[i] - kv_coord[i] >= 0)
                    & (q_coord[i] - kv_coord[i] < kernel_times_dilation)
                    & ((q_coord[i] % dilation[i]) == (kv_coord[i] % dilation[i]))
                )
            else:
                kernel_center_x = q_coord[i].clamp(
                    (kernel_times_dilation - 1) // 2,
                    (image_shape[i] - 1) - (kernel_times_dilation - 1) // 2,
                )
                mask = ((kernel_center_x - kv_coord[i]).abs() <= kernel_times_dilation // 2) & (
                    (q_coord[i] % dilation[i]) == (kv_coord[i] % dilation[i])
                )
```

**Comment by alihassanijr** [comment]
Left a few comments, mostly nits.

**Comment by Leymore** [comment]
Updated. It looks like the wrapper for compiled flex_attention can be removed, while the one for the flex mask cannot. The `create_block_mask` function is not simply compiling something, so the cache is not working.

**Comment by alihassanijr** [comment]
>while the one for the flex mask cannot.

I'm still not convinced about this -- I see virtually no difference between runtimes with and without the wrapper.

Is there a specific case where you observe a difference? If yes, can we unit test it somehow?

**Comment by Leymore** [comment]
I think I was getting the wrong number, sorry about that.

The two key variables are the `functools.lru_cache` and the `_compile` argument from the `create_block_mask` function.

| | w/ functools.lru_cache | w/o functools.lru_cache |
|-|-|-|
| _compile=True | 26.328ms | 26.441ms |
| _compile=False| 26.275ms | 44.568ms |

Numbers are obtained by the following scripts:

```python
import torch
from natten.flex import flex_na2d
import time

batch_size, image_height, image_width, num_head, head_dim = 1, 64, 64, 8, 64
kernel_size, dilation, is_causal = 11, 3, False
query = torch.randn(batch_size, image_height, image_width, num_head, head_dim, device='cuda')
key = torch.randn(batch_size, image_height, image_width, num_head, head_dim, device='cuda')
value = torch.randn(batch_size, image_height, image_width, num_head, head_dim, device='cuda')

# Warm up
for i in range(10):
    output = flex_na2d(query, key, value, kernel_size=kernel_size, dilation=dilation, is_causal=is_causal)

# Test
torch.cuda.synchronize()
time_start = time.time()
for i in range(100):
    output = flex_na2d(query, key, value, kernel_size=kernel_size, dilation=dilation, is_causal=is_causal)
torch.cuda.synchronize()
time_end = time.time()

print(f'Time cost: {(time_end - time_start) / 100 * 1000:.3f}ms')
```

In that case, I removed the `functools.lru_cache` decorator.

I don't think we need to make this a unit test.
